doc/lb.1
.TH lb 1 "September 2005"  "Version 1.0.0" "1.0.0"
.SH NAME
lb \- reads budget file and GnuCash cash-flow reports and merges data to standard output
.SH SYNOPSIS

.BR "lb "
.RB [ -b 
budget.csv ... ] 
.RB [ -r
report.html ... ]
.RB [ -s 
NN.NN MM.MM ... ] 
.RB [ -q ]

.SH DESCRIPTION

.B lb,
The Little Budgeting Tool for GnuCash is a command-line utility
that compares a comma-separated values budget definition with a HTML
cash-flow report from GnuCash.  It merges the two documents into a
report that shows (for each account) the budgeted amount, the actual
amount spent, and the difference between the two.

The report is generated based on settings found in your .lbrc file.
See
.BR lbrc (1)
For information on setting up this file (as well as the default
settings).

.SH "OPTIONS"
.TP 8
.BI -b " budget.csv ..."
The input budget definition files (separated by space).
.TP 8
.BI -r " report.html ..."
The input report file, a cash-flow report generated by 
.B GnuCash.
.TP 8
.BI -s " NN.NN MM.MM ..."
The scale factor to scale the budget files.  Every value in the budget
file will be multipled by this number.  The first budget corresponds
to the first scale value given, second budget - second value, etc.
This is a space separated list.
.TP 8
.BI -q
Quiet operation -- no title or version number.

.SH "BUDGET.CSV"
A budget definition file is a comma separated values file that has
your budget defined in it.  The format is pretty simple.  Comments are
lines that begin with a hash (#).  Account name on the left
(e.g. Expenses:Dining), then a comma (,), then the amount to budget
for that account (e.g. 40.00).  A very basic budget file might look
like this:

.PD 0
# budget.csv
.P
#
.P
# Monthly budget
.P
Income:Salary,1000.00
.P
Expenses:Tithe,100.00
.P
Assets:Current Assets:Savings Account,100.00
.P
Expenses:Rent,250.00
.P
Expenses:Grocery,250.00
.P
Expenses:Auto:Gas,80.00
.P
Expenses:Utilities:Electric,80.00
.P
Expenses:Misc,140.00
.PD

This file is located in the example directory.  Notice that the first
line is the income, and the sum of all the other values is exactly the
same as the income.  If you don't already know, this is good practice
in budgeting -- budget every penny.

If you wish to maintain this file with a spreadsheet application (like
OpenOffice.org or Gnumeric), watch out for:

1. Blank lines in the data.  You will end up exporting a line with
just a comma.  This confuses the program, and you'll get unexpected
results.  Prepend all blank lines with a hash(#).

2. Don't export the CSV with quotation marks (").  This is typically
the default for most spreadsheets.  If it puts quotation marks around
your account names and values, you'll get unexpected results.

Some tricks:

1. If you run 'lb -b budget.csv' it will automatically give you a
bottom-line balance.  For the example file, it will be 0.00.  This a
good check to make sure you haven't over-extended your budget.

2. If you want to run a "percentage" budget, you can do this as well.
Say that you call your income 1.00.  Then each budgeted
item will be a percentage of that.  (E.g. Expenses:Grocery,.25 for
25% going toward grocery.)  When you run the report, you can scale it
by a factor.  To show what the budget is for $2148.62 of income, run
the
report as 'lb -b budget.csv -s 2148.62'  All of the accounts will be
scaled accordingly.

3. To save the report, just redirect the output.  E.g. 'lb -b
budget.csv -r report.html > summary.txt'

.SH "REPORT"
You can't generate a report by hand.  This tool was designed to use
with GnuCash reports.  So, I'll step you through how to do this with
GnuCash 1.8.9.  Your version will no doubt be similar.

From the accounts screen (the main window when you open GnuCash),
select from the top menu Reports | Income & Expense | Cash Flow.  By
default it will give you a summary for all the account for the entire
year.  However, you probably want to get a report for last month.  Or
for a 4-week period corresponding to your paycheck.  Here's how to
tweak it for what you need.

1. Click the Options button.

2. Under the General tab, select the start and end dates for the
period you are interested in.  (E.g. From: Start of previous month,
To: End of previous month.)

3. Under the Accounts tab, select the account that you want to study.
By default, it has all the Assets accounts selected.  However, for
most personal budgeting, you may want to just select the
"Assets:Current Assets:Checking Account."  This will track everything
going into and out of that account.  To unselect the other accounts,
just click on them.

4. Click OK.

5. Click the Export button.

6. In the "Save" dialog box, give the name and location for where you
want to save the report.  It's recommended to include the .html
extension.  (E.g. report.html)

You're done.  You can run an LB report with this.

.SH "RUNNING AN LB REPORT"

From the command line, you enter:

$ lb -b budget.csv -r report.html

And you'll get a report that merges the data from each.  If you use
the budget & report from the example directory, your output will look
like this:

.PD 0
$ lb -b budget.csv -r report.html
.P
LB - The Little Budget Tool, (C)2004 by G. M. Beddingfield, Romans
11:36
.P
Beginning date: 10/01/04.
.P
Ending date:    10/31/04.
.P

.P
                              Account    Planned    Actual      Diff
.P
=====================================  ========= ========= =========
.P
                        Income:Salary    1000.00   1048.16    -48.16
.P
                                       ========= ========= =========
.P
                        Total Income:    1000.00   1048.16     48.16
.P

.P
Assets:Current Assets:Savings Account     100.00    100.00      0.00
.P
                                       ========= ========= =========
.P
                        Total Assets:     100.00    100.00      0.00
.P

.P
                    Expenses:Auto:Gas      80.00     70.20      9.80
.P
                     Expenses:Grocery     250.00    275.41    -25.41
.P
                        Expenses:Misc     140.00    136.25      3.75
.P
                        Expenses:Rent     250.00    249.00      1.00
.P
                       Expenses:Tithe     100.00    105.00     -5.00
.P
          Expenses:Utilities:Electric      80.00     74.18      5.82
.P
                                       ========= ========= =========
.P
                      Total Expenses:     900.00    910.04     10.04
.P

.P

.P
                         BOTTOM LINE:       0.00     38.12     38.12
.P

.P
$
.PD

The first two columns, you can get from your respective reports.  It's
the column on the far right that you're probably most interested in.
Negative values (e.g. -25.41 for Expenses:Grocery) mean that you've
overspent.  Positive values mean that you still have money left to
spend in that account (according to the budget).  However, some of
those accounts will have to make up for overspending on Groceries this
month.

.SH YTD

The report will also suply a set of YTD (or Year-To-Date columns).  If
you just run one budget and one report, the YTD columns will look
exactly like the regular report columns.  However, if you specify more
than one budget and/or report, it will consider the first
one 'current' and all the others 'previous'.  The 'previous' ones will
all be added up, including the current, and give you a long-term
status report.

For example:
.P
lb -b mar.csv feb.csv jan.csv -r mar.html feb.html jan.html

Will give you a report for March, and also a YTD report that sums the
values for January, February, and March.

.SH "SEE ALSO"

.BR gnucash (1),
.BR lbgui (1)

.SH AUTHOR

G. M. Beddingfield <gabriel@teuton.org>


ChangeLog
ChangeLog for lb
	
2005-10-28 Gabriel M. Beddingfield <gabriel@teuton.org>
	* Release:  4.0.1
	* Fixed build errors on gcc4 / FC4
	* Updated autoconf tools.


src/core/lb.cpp
/***************************************************************************
 *                                                                         *
 *   LB - The Little Budget Utility for GnuCash                            *
 *                                                                         *
 *   Copyright (C) 2004,2005 by Gabriel M. Beddingfield                    *
 *   gabriel@teuton.org                                                    *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/


/* lb.h
   Gabriel M. Beddingfield
   29MAR2005

   Application definition for lb.
   For merging a gnucash report
   with a published budget.
*/

#include <iostream>
#include <iomanip>
#include <string>
#include <sstream>
#include <stdexcept>
#include <cstdlib> // for atof()
#include <cassert>
#include "lb.h"
#include "lbrc.h"
#include "bud.h"
#include "budget.h"
#include "report.h"

using namespace std;

#ifndef VERSION
#define VERSION "1.0.1"
#endif


lb::lb(void) : _scale(0,0)
{
	assert( _scale.size()==0 );
	parse_args(0,NULL);
	_rc = new lbrc;
}

lb::lb(int argc, const char* argv[]) : _scale(0,0)
{
	assert( _scale.size()==0 );
	parse_args(argc,argv);
	_rc = new lbrc;
}

lb::~lb()
{
	delete _rc;
}

string lb::about(void) const
{
	ostringstream os;

	os << "LB - The Little Budget Tool " << VERSION << 
		" (C)2004,2005 by G. M. Beddingfield, Romans 11:36";

	return os.str();
}

string lb::usage(void) const
{
	ostringstream os;

	os << "  lb [options] -b budget.txt -r report.html" << endl;
	os << endl;
	os << "  Options:" << endl;
	os << endl;
	os << "  -b budget1.txt budget2.txt ... -- Text file with the" << endl;
	os << "          budget figures." << endl;
	os << endl;
	os << "  -r report1.html report2.html ... -- Cash Flow reports" << endl;
	os << "          from Gnucash." << endl;
	os << endl;
	os << "  -s x.xx ... -- Scale factor for budgets.  All" << endl;
	os << "          figures will be multiplied by N." << endl;
	os << endl;
	os << "  -q -- quiet operation.  No display of title or version.";

	return os.str();
}

int lb::report_accounts( accounts& accts, ostream& os ) {

	lbrc::iterator rci;
	accounts::iterator cur;

	os << _rc->header( &accts );

	for( rci=_rc->begin() ; rci!=_rc->end() ; ++rci ) {
		for( cur=accts.begin() ; cur!=accts.end() ; ++cur ) {
			os << rci->oneline( cur );
		}
		os << rci->summary( &accts );
	}

	os << _rc->footer( &accts );
	return 0;
}

/**************************************************************************************
 * csv2vector()
 * Takes a comma-separated string from x and
 * feeds it into an array of strings.
 * Really only for use in parse_args()
 *************************************************************************************/
void lb::csv2vector(_FileT& f, const char* x)
{
	string s(x);
	size_t pl, pr;

	pl = 0;
	pr = s.find(',');
	while(pr != string::npos) {
		f.push_back(s.substr(pl,(pr-pl)));
		pl = pr+1;
		pr = s.find(',',pl);
	}
	f.push_back( s.substr(pl) );

	return;
}

/**************************************************************************************
 * argv2vector()
 * Takes a subset of the argv's and
 * feeds it into an array of strings.
 * Really only for use in parse_args()
 *************************************************************************************/
void lb::argv2vector(_FileT& f, const char* argv[], int i, int in)
{
	int nd = i+in;
	int j;
	_FileT::iterator k;

	for( j=i+1 ; j<=nd ; j++ ) {
		f.push_back( string(argv[j]) );
	}

	return;
}

/**************************************************************************************
 * argv2floatvector()
 * Takes a subset of the argv's and
 * feeds it into an array of strings.
 * Really only for use in parse_args()
 *************************************************************************************/
void lb::argv2floatvector( _ScaleT& f, const char* argv[], int i, int in)
{
	int nd = i+in;
	int j;

	f.erase(f.begin(),f.end());

	for( j=i+1 ; j<=nd ; j++ ) {
		f.push_back( atof(argv[j]) );
	}

	return;
}

/**************************************************************************************
 * csv2flra()
 * Takes a comma-separated string from x and
 * feeds it into an array of floats.
 * Really only for use in parse_args()
 *************************************************************************************/
void lb::csv2flra(vector<double>& f, const char* x)
{
	string s(x),sub;
	size_t pl, pr;

	pl = 0;
	pr = s.find(',');
	while(pr != string::npos) {
		sub=s.substr(pl,(pr-pl));
		f.push_back( atof(sub.c_str()) );
		pl = pr+1;
		pr = s.find(',',pl);
	}
	sub = s.substr(pl);
	f.push_back( atof(sub.c_str()) );

	return;
}

/******************************************************************************************
 * lb::set_args1()
 *
 * This function is used to process options that require no argument.  For example,
 * if the user passes lb '-q', then we need to parse (and mark parsed) just that we
 * need to be quiet
 *****************************************************************************************/
void lb::set_args1( const int i, int& o, int* f, const int argc, const char* argv[] = NULL )
{
	// calling function should not call use to parse args that have
	// already been parsed.  This checks for that.
	if( f[i] ) {
		_bad = true;
	} else {
		o = 1;
		f[i] = 1;
	}
	return;
}

/******************************************************************************************
 * lb::set_args2()
 *
 * This function is used to process options that require a single argument.  For example,
 * if the user passes lb '-s 1', then we need to parse (and mark parsed) both arguments
 * at one time.
 *****************************************************************************************/
void lb::set_args2( const int i, int& o, int* f, const int argc, const char* argv[] = NULL )
{
	// calling function should not call use to parse args that have
	// already been parsed.  This checks for that.
	if( f[i] ) {
		_bad = true;
		return;
	}

	f[i] = 1;
	o = i+1;
	if( i<argc ) {
		if( f[o] ) {
			_bad = true;
		} else {
			f[o] = 1;
		}
	} else {
		o=i;
		_bad = true;
	}

	if( !argv || (argv[o][0]=='-') ) _bad = true;

	return;
}

/******************************************************************************************
 * lb::set_argsn()
 *
 * This function is used to process options that require an indefinate number of arguments.
 * For example, if the user passes lb '-r foo.html bar.html bat.html', then we need to 
 * parse (and mark parsed) all arguments at one time.
 *
 * Returns: (int) number of arguments to process.
 *                                    Note:  Having a return value is a bit of a hack since
 *                                    I already established an API of returning void.  But
 *                                    there was a critical piece of info I needed back.
 *
 * Arguments:  `i`    - (const) The argv[i] index of the current command-line argument
 *                      that is being processed.
 *             `o`    - (volatile) The option flag for parse_args.  It will be set to i
 *             `f`    - (volatile) The int array of flags to show when the argument has
 *                      been parsed, processed, and tagged.  For example;
 *                         lb -b foo.txt -r bar.html
 *                      '-b' has i=1, 'foo.txt' has i=2, etc.  When we parse -b, we need
 *                      to flag parse_args() to *not* fool with i=2 (foo.txt).
 *             `argc` - The argc passed us by the main calling function (whether main or lb)
 *             `argv` - The argv passed us by the main calling function...
 *****************************************************************************************/
int lb::set_argsn( const int i, int& o, int* f, const int argc, const char* argv[] = NULL )
{
	// calling function should not call use to parse args that have
	// already been parsed.  This checks for that.
	if( f[i] ) {
		_bad = true;
		return 0;
	}

	f[i] = 1;
	o = i;

	int j=i+1;
	while( (j<argc) && !f[j] && (argv[j][0] != '-') ) {   // '-' signals an option
		f[j] = 1;
		++j;
	}

	return j-i-1;
}


/******************************************************************************************
 * lb::parse_args()
 *
 * This function initializes the application by setting the options in the class's data.
 *****************************************************************************************/
void lb::parse_args(int argc, const char* argv[])
{
	int b=0, r=0, s=0, q=0;
	int bn=0, rn=0, sn=0;
	_bad = false;
	_quiet = false;
	_budget.resize(0);
	_report.resize(0);

	// parse the command line
	int chk[argc];
	unsigned int i;
	chk[0] = 1;
	assert( argc >= 0 );
	for( i=1; i<(unsigned int)argc; ++i ) chk[i] = 0;

	for( i=1; i<(unsigned int)argc; ++i) {
		if( chk[i] ) continue;
		if(0==strcmp(argv[i],"-b")) { 
			bn = set_argsn( i, b, chk, argc, argv );
			continue;
		}
		if(0==strcmp(argv[i],"-r")) { 
			rn = set_argsn( i, r, chk, argc, argv );
			continue;
		}
		if(0==strcmp(argv[i],"-s")) { 
			sn = set_argsn( i, s, chk, argc, argv );
			continue;
		}
		if(0==strcmp(argv[i],"-q")) {
			set_args1( i, q, chk, argc, argv );
			continue;
		}
	}

	// Warning:  Using direct user input.
	_quiet = q;
	if(b) argv2vector(_budget,argv,b,bn);
	if(r) argv2vector(_report,argv,r,rn);
	if(s) argv2floatvector(_scale,argv,s,sn);

	// make sure there's a _scale[i] for each _budget[i]
	if( 0==_scale.size() ) _scale.push_back(1.0);
	for( i=_scale.size() ; i<_budget.size() ; ++i ) {
		_scale.push_back( _scale[i-1] );
	}


	// Check for bad inputs.
	if((b>=argc)||(r>=argc)||(s>=argc)||!(b||r))
		_bad = true;

	return;
}

int lb::run(int argc, const char* argv[], ostream& os)
{
	parse_args(argc,argv);
	return run(os);
}

int lb::run(ostream& os)
{

	// Standard checks.
	if(!_quiet) os << about() << endl;

	if(_bad) {
		os << endl << usage() << endl;
		return EXIT_FAILURE;
	}

	_rc->reset();
	accounts accts = parse_reports( &os );

	return report_accounts(accts,os);
}

accounts lb::parse_reports( int argc, const char* argv[] )
{
	parse_args(argc,argv);
	return parse_reports();
}

accounts lb::parse_reports( ostream* os )
{
	accounts accts;

	unsigned int i;
	for( i=0 ; i<_budget.size() ; ++i ) {
		try {
			ReadBudget bfile(_budget[i]);
  
			while(bfile.more()) {
				if(i) {
					accts[bfile.acct()].ytd_planned(_scale[i]*bfile.val());
				} else {
					accts[bfile.acct()].planned(_scale[i]*bfile.val());
				}
			}
		} catch ( runtime_error e ) {
			*os << e.what() << endl;
		}
	}

	for( i=0 ; i<_report.size() ; ++i ) {
		try {
			ReadReport rfile(_report[i]);

			if( (i==0)&&(os!=0) ) {
				*os << "Beginning date: " << rfile.beginning() << "." << endl;
				*os << "Ending date:    " << rfile.ending() << "." << endl;
			}

			while(rfile.more()) {
				if( i ) {
					accts[rfile.acct()].ytd_actual(rfile.val());
				} else {
					accts[rfile.acct()].actual(rfile.val());
				}
			}
		} catch ( runtime_error e ) {
			*os << e.what() << endl;
		}
	}


	return accts;
}
